In [2]:
# Manipulación de matrices
import pandas as pd
# Operaciones numéricas
import numpy as np

El set de datos de **nombres** contiene los registros de la cantidad de personas con un nombre específico, nacidas por año en cada estado de Estados Unidos, desde 1910 hasta 2017.

Pandas permite leer diversos tipos de archivo, como csv, json, xlsx, etc. Usualmente, los datos a utilizar en inteligencia artificial se almacenan en formato csv.

In [3]:
# Leer el archivo nombres.csv y convertirlo en un df
df_base = pd.read_csv('nombres.csv')
# Ver los primeros registros del df
df_base.head()

state sex  year     name  quantity
0    MO   F  1910     Mary       611
1    MO   F  1910    Helen       313
2    MO   F  1910  Dorothy       270
3    MO   F  1910  Mildred       267
4    MO   F  1910     Ruth       237

In [8]:
# Ver los primeros registros del df
df_base.head(100)

state sex  year        name  quantity
0     MO   F  1910        Mary       611
1     MO   F  1910       Helen       313
2     MO   F  1910     Dorothy       270
3     MO   F  1910     Mildred       267
4     MO   F  1910        Ruth       237
5     MO   F  1910       Marie       218
6     MO   F  1910    Margaret       207
7     MO   F  1910      Gladys       167
8     MO   F  1910        Edna       165
9     MO   F  1910        Anna       162
10    MO   F  1910       Hazel       143
11    MO   F  1910     Frances       133
12    MO   F  1910       Grace       132
13    MO   F  1910     Lucille       129
14    MO   F  1910      Thelma       129
15    MO   F  1910    Virginia       128
16    MO   F  1910       Alice       126
17    MO   F  1910        Ruby       119
18    MO   F  1910        Opal       118
19    MO   F  1910       Ethel       116
20    MO   F  1910       Irene       104
21    MO   F  1910     Lillian       104
22    MO   F  1910       Mabel       103
23    MO   F  1910    Florence       100
24    MO   F  1910      Louise        99
25    MO   F  1910       Clara        98
26    MO   F  1910       Elsie        94
27    MO   F  1910   Elizabeth        93
28    MO   F  1910      Martha        93
29    MO   F  1910      Evelyn        90
30    MO   F  1910     Bernice        87
31    MO   F  1910       Edith        87
32    MO   F  1910   Catherine        85
33    MO   F  1910     Pauline        85
34    MO   F  1910      Bessie        84
35    MO   F  1910      Bertha        83
36    MO   F  1910        Rose        81
37    MO   F  1910    Gertrude        77
38    MO   F  1910      Myrtle        77
39    MO   F  1910      Nellie        77
40    MO   F  1910         Eva        75
41    MO   F  1910       Pearl        75
42    MO   F  1910      Beulah        69
43    MO   F  1910        Lena        68
44    MO   F  1910      Esther        67
45    MO   F  1910   Josephine        66
46    MO   F  1910         Ida        64
47    MO   F  1910        Alma        62
48    MO   F  1910       Velma        58
49    MO   F  1910        Emma        57
50    MO   F  1910     Georgia        56
51    MO   F  1910       Viola        55
52    MO   F  1910       Laura        53
53    MO   F  1910  Marguerite        53
54    MO   F  1910      Minnie        53
55    MO   F  1910       Agnes        51
56    MO   F  1910       Leona        51
57    MO   F  1910     Blanche        50
58    MO   F  1910        Ella        50
59    MO   F  1910     Juanita        49
60    MO   F  1910   Katherine        48
61    MO   F  1910        Vera        48
62    MO   F  1910       Wilma        48
63    MO   F  1910      Goldie        47
64    MO   F  1910      Jessie        46
65    MO   F  1910      Violet        46
66    MO   F  1910    Marjorie        45
67    MO   F  1910       Betty        43
68    MO   F  1910     Eleanor        40
69    MO   F  1910      Stella        40
70    MO   F  1910     Alberta        39
71    MO   F  1910        Fern        39
72    MO   F  1910       Julia        39
73    MO   F  1910       Sarah        39
74    MO   F  1910         Ada        38
75    MO   F  1910    Beatrice        38
76    MO   F  1910        Lois        37
77    MO   F  1910         Ann        36
78    MO   F  1910      Lillie        36
79    MO   F  1910         Mae        36
80    MO   F  1910      Sylvia        36
81    MO   F  1910        Lela        35
82    MO   F  1910      Vivian        35
83    MO   F  1910     Loretta        34
84    MO   F  1910       Mable        34
85    MO   F  1910        Alta        32
86    MO   F  1910        Dora        32
87    MO   F  1910       Doris        32
88    MO   F  1910      Geneva        32
89    MO   F  1910        Lola        32
90    MO   F  1910      Jewell        31
91    MO   F  1910      Lorene        31
92    MO   F  1910       Ellen        30
93    MO   F  1910        Erma        30
94    MO   F  1910   Genevieve        30
95    MO   F  1910     Kathryn        30
96    MO   F  1910      Eu

In [9]:
# Mostrar información sobre el df creado
df_base.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5933561 entries, 0 to 5933560
Data columns (total 5 columns):
 #   Column    Dtype 
---  ------    ----- 
 0   state     object
 1   sex       object
 2   year      int64 
 3   name      object
 4   quantity  int64 
dtypes: int64(2), object(3)
memory usage: 226.3+ MB


## Encontrando a Mary

El primer registro muestra que en el estado `MO` nacieron 611 **mujeres** llamadas _Mary_ durante el año 1910. Ante esto podríamos plantearnos las siguientes interrogantes:

- ¿Hay hombres llamados Mary?
- ¿Cuál es la cantidad de hombres llamados Mary en cada estado?
- ¿Cuál es la cantidad promedio de hombres llamados Mary en cada estado?
- ¿Cuáles son los tres estados donde hay más hombres llamados Mary?

Vamos a resolver dichas interrogantes en el siguiente ejemplo.

Lo primero será obtener todos los registros de hombres llamados Mary en todo el conjunto de datos.

In [10]:
# Las condiciones se definen como muestra el ejemplo
# Almacenamos el resultado del filtro en una nueva variable llamada df_mary
df_mary = df_base[
    (df_base['sex'] == 'M') &
    (df_base['name'] == 'Mary')
]
# Consultamos los primeros cinco registros
df_mary.head()

state sex  year  name  quantity
76848    MO   M  1913  Mary         6
78179    MO   M  1916  Mary         5
79914    MO   M  1920  Mary         9
80430    MO   M  1921  Mary         7
81976    MO   M  1924  Mary         5

Los datos de `sex` y `name` ya no son relevantes para este set de datos, pues sabemos que todos son `M` y `Mary` respectivamente. Además, ya que no tendremos en cuenta el año para el análisis, tampoco vamos a necesitarlo.

Ya que únicamente necesitamos dos datos para el análisis, podemos crear otro df que contenga únicamente esos dos datos.

In [11]:
# Generar un nuevo df
nuevo_df = df_mary[['state', 'quantity']]
nuevo_df.head()

state  quantity
76848    MO         6
78179    MO         5
79914    MO         9
80430    MO         7
81976    MO         5

### Bonus:

Si quisiéramos saber cuántos hombres Mary han habido en total, podemos generar la sumatoria de forma similar a la estadística descriptiva.

In [12]:
nuevo_df[['quantity']].sum()

quantity    9638
dtype: int64

Ahora que tenemos el set de datos que necesitamos, podemos agrupar la información para realizar el análisis propuesto. Los grupos en un df se crean con el método `groupby`, indicando la serie por la que se desea agrupar.

Una vez se han agrupado, es necesario establecer una función de agregación, que servirá para establecer la forma de manipular los datos no agrupados. Esta agregación se realiza con el método `agg` de un df agrupado, indicando como parámetro la función a utilizar.

In [13]:
# Agrupar por estado utilizando la función sum
estados_df = nuevo_df.groupby('state').agg('sum')
estados_df.head()

quantity
state          
AL          519
AR          159
AZ            6
CA          540
CO           18

In [14]:
# Consultar todos los registros del nuevo DataFrame
estados_df

quantity
state          
AL          519
AR          159
AZ            6
CA          540
CO           18
CT           26
DC           51
FL          127
GA          482
IA           35
IL          570
IN          213
KS            5
KY          343
LA          237
MA          220
MD           92
MI          293
MN           59
MO          236
MS          392
NC          576
NE           10
NJ          137
NM           69
NY          785
OH          543
OK           92
PA          668
SC          361
TN          316
TX          969
UT            5
VA          263
WA           10
WI           92
WV          119

In [20]:
# Reto:
# Obtener la cantidad promedio de Mary por estado usando estados_df
estados_df = nuevo_df.groupby('state').agg('mean')
estados_df.head()

quantity
state           
AL     10.380000
AR      6.625000
AZ      6.000000
CA      8.852459
CO      6.000000

In [21]:
# Ordenar el set de datos por la mayor cantidad
estados_df.sort_values('quantity', ascending=False)

quantity
state           
TX     13.842857
DC     12.750000
NC     12.521739
NY     11.544118
PA     11.133333
AL     10.380000
IL     10.363636
GA     10.041667
MS      9.333333
TN      9.028571
KY      9.026316
SC      9.025000
OH      8.901639
CA      8.852459
MI      7.918919
VA      7.514286
MO      7.375000
MA      6.875000
LA      6.771429
AR      6.625000
IN      6.454545
FL      6.350000
NM      6.272727
WV      6.263158
NJ      6.227273
MD      6.133333
OK      6.133333
WI      6.133333
CO      6.000000
AZ      6.000000
MN      5.900000
IA      5.833333
CT      5.200000
NE      5.000000
KS      5.000000
UT      5.000000
WA      5.000000

In [25]:
# Reto
# Obtener los 10 registros con la mayor cantidad de Mary
estados_df.sort_values('quantity', ascending=True).head(10)

quantity
state          
WA     5.000000
UT     5.000000
NE     5.000000
KS     5.000000
CT     5.200000
IA     5.833333
MN     5.900000
AZ     6.000000
CO     6.000000
OK     6.133333

Al utilizar una serie para agrupar un df, esa serie se convierte en el nuevo índice.

In [26]:
# Por ello es posible acceder por estado
estados_df.loc['CA']

quantity    8.852459
Name: CA, dtype: float64

In [27]:
# Para convertir el índice en una columna del df, se puede generar uno nuevo aleatorio con reset_index
# Nota al uso de inplace=True para modificar el df
estados_df.reset_index(inplace=True)
estados_df


state   quantity
0     AL  10.380000
1     AR   6.625000
2     AZ   6.000000
3     CA   8.852459
4     CO   6.000000
5     CT   5.200000
6     DC  12.750000
7     FL   6.350000
8     GA  10.041667
9     IA   5.833333
10    IL  10.363636
11    IN   6.454545
12    KS   5.000000
13    KY   9.026316
14    LA   6.771429
15    MA   6.875000
16    MD   6.133333
17    MI   7.918919
18    MN   5.900000
19    MO   7.375000
20    MS   9.333333
21    NC  12.521739
22    NE   5.000000
23    NJ   6.227273
24    NM   6.272727
25    NY  11.544118
26    OH   8.901639
27    OK   6.133333
28    PA  11.133333
29    SC   9.025000
30    TN   9.028571
31    TX  13.842857
32    UT   5.000000
33    VA   7.514286
34    WA   5.000000
35    WI   6.133333
36    WV   6.263158

Finalmente, cualquier df puede ser exportado a diversos formatos. Por ejemplo, exportemos `estados_df` a un archivo de excel.

In [28]:
estados_df.to_excel('estados_mary.xlsx', index=None)